crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs

```diff
@@ -0,0 +1,800 @@
+            SequencerStateUpdatorError::Shutdown
+        })?;
+        let nonce_when_queued = self
+            .execution_backend
+            .get_current_nonce_for_user(&credential_id);
+        let queued_at = Instant::now();
+        queue_receiver.await.unwrap_or_else(|_| {
+            // The oneshot sender was dropped. This should normally only happen
+            // on shutdown, or if there's a bug.
+            let current_nonce = self
+                .execution_backend
+                .get_current_nonce_for_user(&credential_id);
+            err_invalid_nonce::<S, Rt>(
+                tx_hash,
+                tx_nonce,
+                current_nonce,
+                nonce_when_queued,
+                queued_at,
+                credential_id,
+                InvalidNonceReason::EvictedBeforeExecution,
+            )
+        })
+    }
+
+    pub async fn mark_tx_persisted(&self, credential_id: CredentialId) {
+        let _ = self
+            .buffer_sender_channel
+            .send(NonceBufferInput::TxPersisted { credential_id })
+            .await;
+    }
+}
+
+/// Helper enum for error logging when the nonce queue rejects a transaction.
+enum InvalidNonceReason {
+    /// Nonce is in the past or beyond the queue limit. Tx rejected immediately when received.
+    Invalid,
+    /// Tx was queued but the timeout was reached before pre-requisites were all received, so tx
+    /// was evicted from the queue.
+    Timeout,
+    /// Tx was queued but executor oneshot was dropped. This happens on sequencer shutdown OR if
+    /// the queue enters an inconsistent state and internally evicts stale transactions (which
+    /// should almost never happen in practice unless the queue has a bug).
+    EvictedBeforeExecution,
+    /// The nonce was queued for execution, but a new transaction with the same nonce arrived
+    /// before it could be executed.
+    Replaced,
+}
+
+/// Helper function to create an invalid nonce error
+fn err_invalid_nonce<S: Spec, Rt: Runtime<S>>(
+    tx_hash: TxHash,
+    tx_nonce: u64,
+    expected_nonce: u64,
+    nonce_when_queued: u64,
+    instant_queued: std::time::Instant,
+    credential_id: CredentialId,
+    queue_rejection_reason: InvalidNonceReason,
+) -> TransactionReceiverResult<S, Rt> {
+    let was_queued_msg = format!("The sequencer queued the transaction for reordering {} ms ago, when the user's nonce was {nonce_when_queued}.", instant_queued.elapsed().as_millis());
+    let queue_error_msg = match queue_rejection_reason {
+        InvalidNonceReason::Invalid => "The sequencer did not attempt to queue the transaction as it was not within valid queue limits (either in the past, or beyond the max limit).".to_string(),
+        InvalidNonceReason::Timeout => format!("{was_queued_msg} In that time, the sequencer did not receive all the transactions leading up to this tx's nonce, so it has timed out and is being evicted from the queue."),
```

Comment:
nit: let's change "did not receive" to "did not accept" so that we have flexibility to reject txs even if their prereqs have been enqueued.